charmcraft/charm_builder.py

```diff
@@ -0,0 +1,330 @@
+        dest_hookpath = self.buildpath / HOOKS_DIR
+        if not dest_hookpath.exists():
+            dest_hookpath.mkdir()
+
+        # get those built hooks that we need to replace because they are pointing to the
+        # entrypoint directly and we need to fix the environment in the middle
+        current_hooks_to_replace = []
+        for node in dest_hookpath.iterdir():
+            if node.resolve() == linked_entrypoint:
+                current_hooks_to_replace.append(node)
+                node.unlink()
+                logger.debug(
+                    "Replacing existing hook %r as it's a symlink to the entrypoint",
+                    node.name,
+                )
+
+        # include the mandatory ones and those we need to replace
+        hooknames = MANDATORY_HOOK_NAMES | {x.name for x in current_hooks_to_replace}
+        for hookname in hooknames:
+            logger.debug("Creating the %r hook script pointing to dispatch", hookname)
+            dest_hook = dest_hookpath / hookname
+            if not dest_hook.exists():
+                relative_link = relativise(dest_hook, dispatch_path)
+                dest_hook.symlink_to(relative_link)
+
+    def handle_dependencies(self):
+        """Handle from-directory and virtualenv dependencies."""
+        logger.debug("Installing dependencies")
+
+        # virtualenv with other dependencies (if any)
+        if self.requirement_paths:
+            retcode = _process_run(["pip3", "--version"])
+            if retcode:
+                raise CommandError("problems using pip")
+
+            venvpath = self.buildpath / VENV_DIRNAME
+            cmd = [
+                "pip3",
+                "install",  # base command
+                "--target={}".format(
+                    venvpath
+                ),  # put all the resulting files in that specific dir
+            ]
+            if _pip_needs_system():
+                logger.debug("adding --system to work around pip3 defaulting to --user")
+                cmd.append("--system")
+            for reqspath in self.requirement_paths:
+                cmd.append(
+                    "--requirement={}".format(reqspath)
+                )  # the dependencies file(s)
+            retcode = _process_run(cmd)
+            if retcode:
+                raise CommandError("problems installing dependencies")
+
+
+def _pip_needs_system():
+    """Determine whether pip3 defaults to --user, needing --system to turn it off."""
+    cmd = [
+        "python3",
+        "-c",
+        (
+            "from pip.commands.install import InstallCommand; "
+            'assert InstallCommand().cmd_opts.get_option("--system") is not None'
+        ),
+    ]
+    proc = subprocess.run(cmd, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
+    return proc.returncode == 0
+
+
+def _process_run(cmd, **kwargs) -> int:
+    logger.debug("Running external command %s", cmd)
+    try:
+        ret = subprocess.run(cmd, **kwargs)
```

Comment:
Shouldn't the output of this command be captured and sent to `logger.debug`?

Reply:
We could, but if we just dump it to stdout it will be captured by the parent process' `subprocess.run` (which is charmcraft's `polite_exec`) and it will handle it.

Reply:
This is how the charmcraft log currently looks like:
```
2021-07-20 10:15:38,777  charmcraft.commands.build                DEBUG    :: Creating the dispatch mechanism
2021-07-20 10:15:38,777  charmcraft.commands.build                DEBUG    :: Creating the 'upgrade-charm' hook script pointing to dispatch
2021-07-20 10:15:38,777  charmcraft.commands.build                DEBUG    :: Creating the 'install' hook script pointing to dispatch
2021-07-20 10:15:38,777  charmcraft.commands.build                DEBUG    :: Creating the 'start' hook script pointing to dispatch
2021-07-20 10:15:38,777  charmcraft.commands.build                DEBUG    :: Installing dependencies
2021-07-20 10:15:38,777  charmcraft.commands.build                DEBUG    :: Running external command ['pip3', '--version']
2021-07-20 10:15:39,312  charmcraft.commands.build                DEBUG    :: pip 20.0.2 from /home/claudio/.venv/charmcraft/lib/python3.8/site-packages/pip (python 3.8)
2021-07-20 10:15:39,356  charmcraft.commands.build                DEBUG    :: Running external command ['pip3', 'install', '--target=/home/claudio/charmcraft/charmcraft/testdir/build/venv', '--requirement=/home/claudio/charmcraft/charmcraft/testdir/requirements.txt']
2021-07-20 10:15:40,375  charmcraft.commands.build                DEBUG    :: Processing /home/claudio/.cache/pip/wheels/5c/8e/62/9f88a97f7c322b0b5d69d4ce952e32da0f03132ba92d467fb7/ops-1.2.0-py3-none-any.whl
2021-07-20 10:15:40,461  charmcraft.commands.build                DEBUG    :: Collecting PyYAML
2021-07-20 10:15:40,463  charmcraft.commands.build                DEBUG    ::   Using cached PyYAML-5.4.1-cp38-cp38-manylinux1_x86_64.whl (662 kB)
2021-07-20 10:15:40,608  charmcraft.commands.build                DEBUG    :: Installing collected packages: PyYAML, ops
2021-07-20 10:15:40,679  charmcraft.commands.build                DEBUG    :: Successfully installed PyYAML-5.4.1 ops-1.2.0
```
This is how it would look if we also log this subprocess.run output (too wide and doesn't add valuable information IMHO, but it's your call to decide which one is better for the project):
```
2021-07-20 10:11:36,177  charmcraft.commands.build                DEBUG    :: __main__                       Creating the dispatch mechanism
2021-07-20 10:11:36,177  charmcraft.commands.build                DEBUG    :: __main__                       Creating the 'start' hook script pointing to dispatch
2021-07-20 10:11:36,177  charmcraft.commands.build                DEBUG    :: __main__                       Creating the 'upgrade-charm' hook script pointing to dispatch
2021-07-20 10:11:36,177  charmcraft.commands.build                DEBUG    :: __main__                       Creating the 'install' hook script pointing to dispatch
2021-07-20 10:11:36,177  charmcraft.commands.build                DEBUG    :: __main__                       Installing dependencies
2021-07-20 10:11:36,178  charmcraft.commands.build                DEBUG    :: __main__                       Running external command ['pip3', '--version']
2021-07-20 10:11:36,715  charmcraft.commands.build                DEBUG    :: __main__                       :: pip 20.0.2 from /home/claudio/.venv/charmcraft/lib/python3.8/site-packages/pip (python 3.8)
2021-07-20 10:11:36,760  charmcraft.commands.build                DEBUG    :: __main__                       Running external command ['pip3', 'install', '--target=/home/claudio/charmcraft/charmcraft/testdir/build/venv', '--requirement=/home/claudio/charmcraft/charmcraft/testdir/requirements.txt']
2021-07-20 10:11:37,645  charmcraft.commands.build                DEBUG    :: __main__                       :: Processing /home/claudio/.cache/pip/wheels/5c/8e/62/9f88a97f7c322b0b5d69d4ce952e32da0f03132ba92d467fb7/ops-1.2.0-py3-none-any.whl
2021-07-20 10:11:37,743  charmcraft.commands.build                DEBUG    :: __main__                       :: Collecting PyYAML
2021-07-20 10:11:37,745  charmcraft.commands.build                DEBUG    :: __main__                       ::   Using cached PyYAML-5.4.1-cp38-cp38-manylinux1_x86_64.whl (662 kB)
2021-07-20 10:11:37,890  charmcraft.commands.build                DEBUG    :: __main__                       :: Installing collected packages: PyYAML, ops
2021-07-20 10:11:37,966  charmcraft.commands.build                DEBUG    :: __main__                       :: Successfully installed PyYAML-5.4.1 ops-1.2.0
```

Reply:
I prefer your first version, but with a slight detail. The problem I still found is that it's confusing what's part of the "first subprocess" (the builder itself) and the "second one" (the pip called from the builder).

What do you think of something like the following:

```
2021-07-20 10:15:38,777  charmcraft.commands.build                DEBUG    :: Creating the dispatch mechanism
2021-07-20 10:15:38,777  charmcraft.commands.build                DEBUG    :: Creating the 'upgrade-charm' hook script pointing to dispatch
2021-07-20 10:15:38,777  charmcraft.commands.build                DEBUG    :: Creating the 'install' hook script pointing to dispatch
2021-07-20 10:15:38,777  charmcraft.commands.build                DEBUG    :: Creating the 'start' hook script pointing to dispatch
2021-07-20 10:15:38,777  charmcraft.commands.build                DEBUG    :: Installing dependencies
2021-07-20 10:15:38,777  charmcraft.commands.build                DEBUG    :: Running external command ['pip3', '--version']
2021-07-20 10:15:39,312  charmcraft.commands.build                DEBUG    ::     :: pip 20.0.2 from /home/claudio/.venv/charmcraft/lib/python3.8/site-packages/pip (python 3.8)
2021-07-20 10:15:39,356  charmcraft.commands.build                DEBUG    ::     :: Running external command ['pip3', 'install', '--target=/home/claudio/charmcraft/charmcraft/testdir/build/venv', '--requirement=/home/claudio/charmcraft/charmcraft/testdir/requirements.txt']
2021-07-20 10:15:40,375  charmcraft.commands.build                DEBUG    ::     :: Processing /home/claudio/.cache/pip/wheels/5c/8e/62/9f88a97f7c322b0b5d69d4ce952e32da0f03132ba92d467fb7/ops-1.2.0-py3-none-any.whl
2021-07-20 10:15:40,461  charmcraft.commands.build                DEBUG    ::     :: Collecting PyYAML
2021-07-20 10:15:40,463  charmcraft.commands.build                DEBUG    ::     ::   Using cached PyYAML-5.4.1-cp38-cp38-manylinux1_x86_64.whl (662 kB)
2021-07-20 10:15:40,608  charmcraft.commands.build                DEBUG    ::     :: Installing collected packages: PyYAML, ops
2021-07-20 10:15:40,679  charmcraft.commands.build                DEBUG    ::     :: Successfully installed PyYAML-5.4.1 ops-1.2.0
```